### ISSUE TITLE
increment_build_number increase version number instead of build number

### ISSUE THREAD

**Issue by darren1982** [fastlane/fastlane#10328]
### New Issue Checklist

- [ x ] Updated fastlane to the latest version
- [ x ] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ x  ] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [ x ] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
I added increment_build_number to my fastfile so that it will increase the build number for every build and it won't upload a build with existing build number to appstore. 

Instead of increasing my build number, it actually increase my version number. 

See the output below. In the step "increment_build_number", it get my current version number "1", instead of build number then increase it and set it to 2. Then at the end of uploading ipa, it says "There already exists a binary upload with build '2' for version '1.0'"" 

Now it seems like treating the increased version number, "2" as build number. 

---------------------------
```

[15:45:38]: ------------------------------------
[15:45:38]: --- Step: increment_build_number ---
[15:45:38]: ------------------------------------
Current version of project whmcs is: 
    1

/Users/shuyih/dev/pgeon-consumer/ios
[15:45:39]: $ cd /Users/shuyih/dev/pgeon-consumer/ios && agvtool next-version -all && cd -
[15:45:40]: ▸ Setting version of project whmcs to:
[15:45:40]: ▸ 2.
[15:45:40]: ▸ Also setting CFBundleVersion key (assuming it exists)
[15:45:40]: ▸ Updating CFBundleVersion in Info.plist(s)...
[15:45:40]: ▸ Updated CFBundleVersion in "whmcs.xcodeproj/../whmcs/Info.plist" to 2
```
....

[Transporter Error Output]: ERROR ITMS-90189: "Redundant Binary Upload. There already exists a binary upload with build '2' for version '1.0'"

##### Complete output when running fastlane, including the stack trace and command used
> You can use: `--capture_output` as the last commandline argument to get that collected for you

### Captured Output

Command Used: `beta`
<details><summary>Output/Log</summary>

+------------------+--------------------------------------+
|                 Summary for gym 2.56.0                  |
+------------------+--------------------------------------+
| workspace        | ./whmcs.xcworkspace                  |
| scheme           | whmcs                                |
| destination      | generic/platform=iOS                 |
| output_name      | whmcs                                |
| build_path       | /Users/shuyih/Library/Developer/Xco  |
|                  | de/Archives/2017-09-14               |
| clean            | false                                |
| output_directory | .                                    |
| silent           | false                                |
| skip_package_ipa | false                                |
| buildlog_path    | ~/Library/Logs/gym                   |
| xcode_path       | /Applications/Xcode.app              |
+------------------+--------------------------------------+

+--------------------------------------+----------------------+
|                   Generated Build Command                   |
+--------------------------------------+----------------------+
| Option                               | Value                |
+--------------------------------------+----------------------+
| set -o pipefail &&                   |                      |
| xcodebuild                           |                      |
| -workspace ./whmcs.xcworkspace       |                      |
| -scheme whmcs                        |                      |
| -destination                         | generic/platform=iOS |
| -archivePath                         |                      |
| /Users/shuyih/Library/Developer/Xco  |                      |
| de/Archives/2017-09-14/whmcs\        |                      |
| 2017-09-14\ 19.28.01.xcarchive       |                      |
| archive                              |                      |
| | tee                                |                      |
| /Users/shuyih/Library/Logs/gym/whmc  |                      |
| s-whmcs.log                          |                      |
| | xcpretty                           |                      |
+--------------------------------------+----------------------+
+--------------------------------------+--------------------------------------+
|                          Generated Package Command                          |
+--------------------------------------+--------------------------------------+
| Option                               | Value                                |
+--------------------------------------+--------------------------------------+
| /usr/bin/xcrun                       |                                      |
| /Users/shuyih/.fastlane/bin/bundle/  |                                      |
| lib/ruby/gems/2.2.0/gems/fastlane-2  |                                      |
| .56.0/gym/lib/assets/wrap_xcodebuil  |                                      |
| d/xcbuild-safe.sh -exportArchive     |                                      |
| -exportOptionsPlist                  | /var/folders/g5/zd87tj1s4zn8x5m5btj  |
|                                      | bhcy80000gn/T/gym_config20170914-95  |
|                                      | 478-dhn1tz.plist                     |
| -archivePath                         |                                      |
| /Users/shuyih/Library/Developer/Xco  |                                      |
| de/Archives/2017-09-14/whmcs\        |                                      |
| 2017-09-14\ 19.28.01.xcarchive       |                                      |
| -exportPath                          | /var/folders/g5/zd87tj1s4zn8x5m5btj  |
|                                      | bhcy80000gn/T/gym_output20170914-95  |
|                                      | 478-1iasjun                          |
+--------------------------------------+--------------------------------------+

Looking for iTunes Connect Team with ID 118706343

+------+------------------------+-------------+
|              fastlane summary               |
+------+------------------------+-------------+
| Step | Action                 | Time (in s) |
+------+------------------------+-------------+
| 1    | Verifying required     | 0           |
|      | fastlane version       |             |
| 2    | default_platform       | 0           |
| 3    | cocoapods              | 4           |
| 4    | increment_build_numbe  | 2           |
|      | r                      |             |
| 5    | gym                    | 161         |
| 💥   | pilot                  | 45          |
+------+------------------------+-------------+

```

</details>

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.12.6                                     |
| Ruby                        | 2.2.4                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.11.0 (Apple Git-81)           |
| Installation Source         | ~/.fastlane/bin/bundle/bin/fastlane         |
| Host                        | Mac OS X 10.12.6 (16G29)                    |
| Ruby Lib Dir                | ~/.fastlane/bin/bundle/lib                  |
| OpenSSL Version             | OpenSSL 1.0.2g  1 Mar 2016                  |
| Is contained                | false                                       |
| Is homebrew                 | true                                        |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 8.3.3                                       |


### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE | en_US.UTF-8 | ✅ |


### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# Customise this file, documentation can be found here:
# https://github.com/fastlane/fastlane/tree/master/fastlane/docs
# All available actions: https://docs.fastlane.tools/actions
# can also be listed using the `fastlane actions` command

# Change the syntax highlighting to Ruby
# All lines starting with a # are ignored when running `fastlane`

# If you want to automatically update fastlane if a new version is available:
# update_fastlane

# This is the minimum version number required.
# Update this, if you use features of a newer version
fastlane_version "2.56.0"

default_platform :ios

platform :ios do
  before_all do
    # ENV["SLACK_URL"] = "https://hooks.slack.com/services/..."
    cocoapods
    
  end

  desc "Runs all the tests"
  lane :test do
    scan
  end

  desc "Submit a new Beta Build to Apple TestFlight"
  desc "This will also make sure the profile is up to date"
  lane :beta do
    increment_build_number
    # match(type: "appstore") # more information: https://codesigning.guide
    gym # Build your app - more options available
    pilot

    # sh "your_script.sh"
    # You can also use other beta testing services here (run `fastlane actions`)
  end

  desc "Deploy a new version to the App Store"
  lane :release do
    # match(type: "appstore")
    # snapshot
    gym # Build your app - more options available
    deliver(force: true)
    # frameit
  end

  # You can define as many lanes as you want

  after_all do |lane|
    # This block is called, only if the executed lane was successful

    # slack(
    #   message: "Successfully deployed new App Update."
    # )
  end

  error do |lane, exception|
    # slack(
    #   message: exception.message,
    #   success: false
    # )
  end
end


# More information about multiple platforms in fastlane: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Platforms.md
# All available actions: https://docs.fastlane.tools/actions

# fastlane reports which actions are used. No personal data is recorded. 
# Learn more at https://github.com/fastlane/fastlane#metrics

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier "xxx" # The bundle identifier of your app
apple_id "xxx" # Your Apple email address

team_id "xxx" # Developer Portal Team ID

# you can even provide different app identifiers, Apple IDs and team names per lane:
# More information: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Appfile.md

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.56.0  | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| slack-notifier            | 1.5.1        |
| CFPropertyList            | 2.3.5        |
| claide                    | 1.0.2        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.5.1        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.6        |
| terminal-notifier         | 1.7.1        |
| unicode-display_width     | 1.1.3        |
| terminal-table            | 1.7.3        |
| plist                     | 3.2.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.1        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.5.0        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| highline                  | 1.7.8        |
| commander-fastlane        | 4.4.5        |
| excon                     | 0.55.0       |
| faraday                   | 0.12.1       |
| unf_ext                   | 0.0.7.4      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20170404 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| fastimage                 | 2.1.0        |
| gh_inspector              | 1.0.3        |
| json                      | 1.8.1        |
| mini_magick               | 4.5.1        |
| multi_json                | 1.12.1       |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.2.1        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 0.0.4        |
| dotenv                    | 2.2.0        |
| bundler                   | 1.14.6       |
| faraday_middleware        | 0.11.0.1     |
| uber                      | 0.0.15       |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 2.1.0        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| little-plugger            | 1.1.4        |
| logging                   | 2.2.2        |
| jwt                       | 1.5.6        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.13.6       |
| libxml-ruby               | 3.0.0        |
| psych                     | 2.0.8        |
</details>

*generated on:* **2017-09-14**
</details>

**Comment by ohayon** [comment]
Hey @darren1982, the `CFBundleVersion` key does in fact point to your build number. The version number itself is assigned to the `CFBundleShortVersionString` see some more docs here: https://developer.apple.com/library/content/technotes/tn2420/_index.html 

Let us know if that doesn't make sense or if there is anything else we can do to help! 🐙  🚀

**Comment by darren1982** [comment]
Hi @ohayon , 

My existing build number is actually 10. I verified from xcode, Info.plist does show as below as 10. 
	<key>CFBundleVersion</key>
	<string>10</string>

However, when I run fastlane beta, increment_build_number, it retrieves my version number 1. Increment it by 1. Then set it to my build number instead. So my build number changed from 10 to become 2. I checked my version number, it increased from 1 to 2 as well. 

So, instead of increment my build number. It increments version number from 1 to 2. Then set it to build number as well. 


[16:19:52]: ------------------------------------
[16:19:52]: --- Step: increment_build_number ---
[16:19:52]: ------------------------------------
Current version of project whmcs is: 
    1

/Users/shuyih/dev/pgeon-consumer/ios
[16:19:54]: $ cd /Users/shuyih/dev/pgeon-consumer/ios && agvtool next-version -all && cd -
[16:19:55]: ▸ Setting version of project whmcs to:
[16:19:55]: ▸ 2.
[16:19:55]: ▸ Also setting CFBundleVersion key (assuming it exists)
[16:19:55]: ▸ Updating CFBundleVersion in Info.plist(s)...
[16:19:56]: ▸ Updated CFBundleVersion in "whmcs.xcodeproj/../whmcs/Info.plist" to 2

**Comment by ohayon** [comment]
Hmm. that is interesting. Is your project set up to use [Apple Generic Versioning](https://developer.apple.com/library/content/qa/qa1827/_index.html)?

**Comment by darren1982** [comment]
Hi @ohayon ,

Yes. My project is using Apple Generic Versioning. I think I figure out why. Once I run agvtool next-version -all 3 times, I checked on the diff of the files that are changed in my repo, I notice that project.pbxproj is changed from "CURRENT_PROJECT_VERSION = 1;” to 
"CURRENT_PROJECT_VERSION = 4”; 

I think that’s the reason why it doesn’t work on first run. It appears that agvtool next-version -all use this CURRENT_PROJECT_VERSION to determine the current build number and then increment it and set it back. Somehow, all the while, I set it through Xcode UI > General > Identity > Build and that doesn’t change this variable. 

So, to resolve this, I manually run agvtool next-version -all  to increment the variable correctly to my current build number 11, then any subsequent fast lane run will succeed. 

Thanks a lot for your pointer.
